### PR TITLE
[oneDPL][test][ranges] + segfault fix for some "ranges" tests

### PR DESCRIPTION
**Pull request by MikeDvorskiy**
Changes in host_all definition to get a sycl::host_accessor instance due to non-placeholder sycl::accessor for the host is deprecated.
It makes  a ten of the range tests as passed.

### CODE REVIEW

File: include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h [hunk 1]
```diff
@@ -118,6 +118,18 @@ struct all_view_fn
     }
 };
 
+#if _ONEDPL_LIBSYCL_VERSION >= 60200
```
[the remark this answers is not in the document]
Comment (MikeDvorskiy) [reply include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h:121]: To void double negative, probably,   `_ONEDPL_SYCL_PLACEHOLDER_HOST_ACCESSOR_DEPRECATED`  is better?

**Review by dmitriy-sobolev — APPROVED** [review]
LGTM.
I confirm that the code works with two version of Intel(R) oneAPI DPC++/C++ Compiler: 
 - `_ONEDPL_LIBSYCL_VERSION == 60200`: segmentation faults are fixed
 - `_ONEDPL_LIBSYCL_VERSION == 60100`: there are no run-time issues and no deprecation warnings related to using a host placeholder accessor.

File: include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h [hunk 2]
```diff
@@ -118,6 +118,18 @@ struct all_view_fn
     }
 };
 
+#if _ONEDPL_SYCL_PLACEHOLDER_HOST_ACCESSOR_DEPRECATED
+struct all_host_view_fn
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h:122]: I compared the public API of `host_all` before and after the change:

  | before | after
-- | -- | --
value_type | + | +
begin | + | +
end | + | +
size | + | +
empty | + | +
require_access | + | -
constructor(buffer, offset, n) | + | -
constructor(accessor) | + | -
noop (shall return the same view) | + | -


Could you make additional changes in order not to break backward compatibility?

File: include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h [hunk 3]
```diff
@@ -130,8 +150,13 @@ inline constexpr all_view_fn<sycl::access::mode::read, __dpl_sycl::__target_devi
 inline constexpr all_view_fn<sycl::access::mode::write, __dpl_sycl::__target_device, sycl::access::placeholder::true_t>
     all_write;
 
+#if _ONEDPL_SYCL_PLACEHOLDER_HOST_ACCESSOR_DEPRECATED
+//CPO to get a sycl::host_accessor instance due to non-placeholder sycl::accessor for the host is deprecated
```
Comment (dmitriy-sobolev) [inline include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h:154]: What do you think about removing that comment? `_ONEDPL_SYCL_PLACEHOLDER_HOST_ACCESSOR_DEPRECATED` macro name is pretty descriptive on its own.

Comment (dmitriy-sobolev) [reply include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h:154]: The comment part about CPO should probably be added prior to: 
```
    operator()(_R&& __r) const -> decltype(::std::forward<_R>(__r))
    {
        return ::std::forward<_R>(__r);
    }
```

**Comment by dmitriy-sobolev** [comment]
The following tests started failing after [[oneDPL][ranges] + 'n' and 'offset' for CPO host_all](https://github.com/oneapi-src/oneDPL/pull/1021/commits/c372929772f63ffaeed1bd9eec49605a6720f1fa):
```
26/47 Test #133: replace_copy_ranges_sycl.pass ...............***Failed    0.18 sec
error at oneDPL/test/parallel_api/ranges/replace_copy_ranges_sycl.pass.cpp:46 - wrong effect from replace_copy, expected sequence of size 10 got sequence of size 0

      Start 135: reverse_copy_ranges_sycl.pass
28/47 Test #135: reverse_copy_ranges_sycl.pass ...............***Failed    0.21 sec
error at oneDPL/test/parallel_api/ranges/reverse_copy_ranges_sycl.pass.cpp:44 - wrong effect from reverse_copy, expected sequence of size 10 got sequence of size 0

      Start 136: reverse_ranges_sycl.pass
29/47 Test #136: reverse_ranges_sycl.pass ....................***Failed    0.23 sec

error at oneDPL/test/parallel_api/ranges/reverse_ranges_sycl.pass.cpp:48 - wrong effect from reverse, expected sequence of size 10 got sequence of size 0

      Start 137: rotate_copy_ranges_sycl.pass
30/47 Test #137: rotate_copy_ranges_sycl.pass ................***Failed    0.28 sec
error at oneDPL/test/parallel_api/ranges/rotate_copy_ranges_sycl.pass.cpp:45 - wrong effect from rotate_copy, expected sequence of size 10 got sequence of size 0
```

**Comment by MikeDvorskiy** [comment]
> I guess you should add this logic from `all_view::__create_accessor`:
> 
> ```
> auto __n_buf = __dpl_sycl::__get_buffer_size(__buf);
> auto __n_acc = (__n > 0 ? __n : __n_buf);
> ```

Probably, yes.... Thanks

File: include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h [hunk 4]
```diff
@@ -118,6 +121,28 @@ struct all_view_fn
     }
 };
 
+#if _ONEDPL_SYCL_PLACEHOLDER_HOST_ACCESSOR_DEPRECATED
+struct all_host_view_fn
+{
+    //An overalod for sycl::buffer template type
```
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h:127]:
```suggestion
    // An overload for sycl::buffer template type
```

File: include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h [hunk 5]
```diff
@@ -118,6 +121,28 @@ struct all_view_fn
...
+    template <typename _T>
+    auto
+    operator()(sycl::buffer<_T, 1> __buf, typename ::std::iterator_traits<_T*>::difference_type __offset = 0,
+               typename ::std::iterator_traits<_T*>::difference_type __n = 0) const
+    {
+        return __internal::__create_accessor<sycl::host_accessor<_T>>(__buf, __offset, __n);
+    }
+
+    //"No operation" overalod for another ranges/views
```
Comment (julianmi) [inline include/oneapi/dpl/pstl/hetero/dpcpp/utils_ranges_sycl.h:136]:
```suggestion
    // "No operation" overload for another ranges/views
```

**Review by julianmi — APPROVED** [review]
LGTM.

**Review by julianmi — APPROVED** [review]

**Pull request merged**